url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/core/QuartzSchedulerThread.java

```diff
@@ -0,0 +1,476 @@
+      this.setContextClassLoader(Thread.currentThread().getContextClassLoader());
+    }
+
+    this.setPriority(threadPrio);
+
+    // start the underlying thread, but put this object into the 'paused'
+    // state
+    // so processing doesn't start yet...
+    paused = true;
+    halted = new AtomicBoolean(false);
+  }
+
+  /*
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Interface.
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
+
+  private long getRandomizedIdleWaitTime() {
+
+    return idleWaitTime - random.nextInt(idleWaitVariablness);
+  }
+
+  /**
+   * <p>
+   * Signals the main processing loop to pause at the next possible point.
+   * </p>
+   */
+  public void togglePause(boolean pause) {
+
+    synchronized (sigLock) {
+      paused = pause;
+
+      if (paused) {
+        signalSchedulingChange(0);
+      } else {
+        sigLock.notifyAll();
+      }
+    }
+  }
+
+  /**
+   * <p>
+   * Signals the main processing loop to pause at the next possible point.
+   * </p>
+   */
+  public void halt() {
+
+    synchronized (sigLock) {
+      halted.set(true);
+
+      if (paused) {
+        sigLock.notifyAll();
+      } else {
+        signalSchedulingChange(0);
+      }
+    }
+  }
+
+  public boolean isPaused() {
+
+    return paused;
+  }
+
+  /**
+   * <p>
+   * Signals the main processing loop that a change in scheduling has been made - in order to interrupt any sleeping that may be occurring while
+   * waiting for the fire time to arrive.
+   * </p>
+   *
+   * @param candidateNewNextFireTime the time (in millis) when the newly scheduled trigger will fire. If this method is being called do to some other
+   *        even (rather than scheduling a trigger), the caller should pass zero (0).
+   */
+  void signalSchedulingChange(long candidateNewNextFireTime) {
+
+    synchronized (sigLock) {
+      signaled = true;
+      signaledNextFireTime = candidateNewNextFireTime;
+      sigLock.notifyAll();
+    }
+  }
+
+  private void clearSignaledSchedulingChange() {
+
+    synchronized (sigLock) {
+      signaled = false;
+      signaledNextFireTime = 0;
+    }
+  }
+
+  public boolean isScheduleChanged() {
+
+    synchronized (sigLock) {
+      return signaled;
+    }
+  }
+
+  public long getSignaledNextFireTime() {
+
+    synchronized (sigLock) {
+      return signaledNextFireTime;
+    }
+  }
+
+  /**
+   * <p>
+   * The main processing loop of the <code>QuartzSchedulerThread</code>.
+   * </p>
+   */
+  @Override
+  public void run() {
+
+    boolean lastAcquireFailed = false;
+
+    while (!halted.get()) {
+      try {
+        // check if we're supposed to pause...
+        synchronized (sigLock) {
+          while (paused && !halted.get()) {
+            try {
+              // wait until togglePause(false) is called...
+              sigLock.wait(1000L);
+            } catch (InterruptedException ignore) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)